### ISSUE TITLE
Calico (TPR->CRD) 2.5.0+ migration does not produce new CRD-based policies

### ISSUE THREAD

**Issue by joshrosso** [projectcalico/calico#1620]
## Expected Behavior
I'm attempting to upgrade a `1.7.5` Kubernetes cluster using TPR-based `SystemNetworkPolicy` to express NetworkPolicies to the new CRD format so that I can upgrade calico (running as canal (flannel+calico)).

I'm following the steps described at https://github.com/projectcalico/calico/tree/master/upgrade/v2.5.

The current set of policies are

```
$ kubectl get systemnetworkpolicy -n kube-system
NAME                                
1-tier                             
2-tier               
egress-allow-to-out       
egress-allow-to-in         
egress-deny-to-dbs      
egress-deny-to-nodes    
egress-implicit-deny    
ingress-deny-from-pubin 
special-collector      
namespace-tools 
some-ingress 
rfc1918      
unlabeled    
3-tier           
```

After running the migration, I'd expect to see a representation of the above in CRD form.

## Current Behavior

The migration job runs and logs can be seen below.

```
$ kubectl logs calico-upgrade-v2.5-qxfz3
Creating Calico CRDs...

customresourcedefinition "globalfelixconfigs.crd.projectcalico.org" created
customresourcedefinition "bgppeers.crd.projectcalico.org" created
customresourcedefinition "ippools.crd.projectcalico.org" created
customresourcedefinition "globalnetworkpolicies.crd.projectcalico.org" created
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
thirdpartyresource "global-config.projectcalico.org" configured
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
thirdpartyresource "ip-pool.projectcalico.org" configured
thirdpartyresource "global-bgp-config.projectcalico.org" created
thirdpartyresource "global-bgp-peer.projectcalico.org" created
Migrating IPPools...
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: 192.2.0.0/16
  spec:
    ipip:
      enabled: true
      mode: always
    nat-outgoing: true
Successfully applied 1 'ipPool' resource(s)
Migrated the IPPools!

Migrating BGPPeers...
[]
No BGPPeers found to migrate.
Moving on...

Migrating GlobalFelixConfig...
apiVersion: v1
items:
- apiVersion: projectcalico.org/v1
  kind: GlobalConfig
  metadata:
    creationTimestamp: 2017-10-06T22:49:12Z
    name: clusterguid
    namespace: kube-system
    resourceVersion: "465"
    selfLink: /apis/projectcalico.org/v1/namespaces/kube-system/globalconfigs/clusterguid
    uid: 9282dfdfd3-aae8-11e7-bbc5-12a5e9a65b38
  spec:
    name: ClusterGUID
    value: 4e3b2b4fd04e4046a92dabbdb1742a95
- apiVersion: projectcalico.org/v1
  kind: GlobalConfig
  metadata:
    creationTimestamp: 2017-10-06T22:49:12Z
    name: clustertype
    namespace: kube-system
    resourceVersion: "464"
    selfLink: /apis/projectcalico.org/v1/namespaces/kube-system/globalconfigs/clustertype
    uid: 928dfdf8-aae8-11e7-bbc5-12a5e9a65b38
  spec:
    name: ClusterType
    value: KDD
- apiVersion: projectcalico.org/v1
  kind: GlobalConfig
  metadata:
    creationTimestamp: 2017-10-06T22:49:12Z
    name: ipinipenabled
    namespace: kube-system
    resourceVersion: "468"
    selfLink: /apis/projectcalico.org/v1/namespaces/kube-system/globalconfigs/ipinipenabled
    uid: 929dfdfdd-aae8-11e7-bbc5-12a5e9a65b38
  spec:
    name: IpInIpEnabled
    value: "true"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

Converted global-config.projectcalico.org/v1 to globalfelixconfigs.crd.projectcalico.org/v1
apiVersion: v1
items:
- apiVersion: crd.projectcalico.org/v1
  kind: GlobalFelixConfig
  metadata:
    name: clusterguid
  spec:
    name: ClusterGUID
    value: 4e3basdfasdfa92dabbdb1742a95
- apiVersion: crd.projectcalico.org/v1
  kind: GlobalFelixConfig
  metadata:
    name: clustertype
  spec:
    name: ClusterType
    value: KDD
- apiVersion: crd.projectcalico.org/v1
  kind: GlobalFelixConfig
  metadata:
    name: ipinipenabled
  spec:
    name: IpInIpEnabled
    value: "true"
kind: List
metadata:
globalfelixconfig "clusterguid" created
globalfelixconfig "clustertype" created
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
globalfelixconfig "ipinipenabled" configured
Migrated the GlobalFelixConfig!

Migrating GlobalBGPConfig...
apiVersion: v1
items: []
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
No GlobalBGPConfig found to migrate.
customresourcedefinition "globalbgpconfigs.crd.projectcalico.org" created
Moving on...

Successfully migrated Calico data!
```

When inspecting the CRDs, I don't see any correlated policies. I'm assuming `globalnetworkpolicies.crd.projectcalico.org` is the place to look, but checking all CRDs related to `crd.projectcalico.org` revealed nothing.

```
$ kubectl get globalnetworkpolicies.crd.projectcalico.org 
No resources found.
```

## Steps to Reproduce (for bugs)
Follow steps at https://github.com/projectcalico/calico/tree/master/upgrade/v2.5.

## Your Environment

* Calico mode: `canal`
* Kubernetes version: `1.7.5`
* Calico version: `quay.io/calico/node:v1.3.0`
* Operating System and version: `Container Linux by CoreOS 1576.5.0 (Ladybug)`

**Comment by caseydavenport** [comment]
@joshrosso thanks for raising.

The `SystemNetworkPolicy` resource was alpha in Calico v2.4 and as you've discovered was removed in Calico v2.5 in favor of `GlobalNetworkPolicy`.

Due to the resource being alpha quality with a few known issues, SNPs were not included in the migration steps you followed. SNPs weren't meant to be generally available and thus weren't documented, etc. It should still be possible to do a manual migration of the data from TPR->CRD via `kubectl` if necessary.

Related to this, I'd note that GlobalNetworkPolicy, while available as a CRD in Calico v2.5, isn't officially supported until Calico v3.0, where first-class documentation and integration into `calicoctl` is available.

**Comment by joshrosso** [comment]
@caseydavenport Thanks for the clarification. 

If I'm understanding correctly, the TPR->CRD migration is really only handling the global settings / configuration of calico (previously held in TPRs) but not the policies themselves.

Would you have any insight as to whether `GlobalNetworkPolicy` is any more/less stable in calico `2.6.1` (what we're initially upgrading to) relative to the existing `SystemNetworkPolicies` we're using in pre-2.5?

**Comment by joshrosso** [comment]
And a third question, if I may.

Would Calico generally advise k8s users to stick to the standard `NetworkPolicy` API where ever possible (ie where the `NetworkPolicy isn't feature deficient) rather than the `GlobalNetworkPolicy` CRD?

**Comment by caseydavenport** [comment]
> Would you have any insight as to whether GlobalNetworkPolicy is any more/less stable in calico 2.6.1 (what we're initially upgrading to) relative to the existing SystemNetworkPolicies we're using in pre-2.5?

@joshrosso officially, GlobalNetworkPolicy isn't generally available until Calico release v3.0, so I'm afraid I can't guarantee GNP support across upgrades from v2.6, etc.

> Would Calico generally advise k8s users to stick to the standard NetworkPolicy API where ever possible (ie where the NetworkPolicy isn't feature deficient) rather than theGlobalNetworkPolicy` CRD?

This one is a bit more nuanced. NetworkPolicy and GlobalNetworkPolicy overlap in capabilities, but GlobalNetworkPolicy (like its name implies) affects the entire cluster, not just a particular namespace, and also provides the user some extra knobs to turn.

I don't think there's a blanket statement I can give re: when to use one or the other, though I suspect that minus the difference in features, most users will find it preferable to stick with a single API. Which one they pick depends on their requirements, use-case(s), organizational structure, etc. Also, for Calico <v3.0 I'd advise against relying on GNP given its pre-GA status. Hopefully that helps!

**Comment by joshrosso** [comment]
This helps clarify a lot. Thank you.
